File: 3128-right-triangles/3128-right-triangles.java
class Solution {
    public long numberOfRightTriangles(int[][] grid) {
        long result = 0;
        int[] rowMap = new int[grid.length], colMap = new int[grid[0].length];
        createfreqRowCol(grid, rowMap, colMap);
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    int row = rowMap[i], column = colMap[j];
                    result = result + (row - 1) * (column - 1);
                }
            }
        }
        return result;
    }
    
    private void createfreqRowCol(int[][] grid, int[] rowMap, int[] colMap) {
        for (int i = 0; i < grid.length; i++) {
            int count = 0;
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    rowMap[i]++;
                    colMap[j]++;
                }
            }
        }
    }

}

File: 3128-right-triangles/3128-right-triangles.cpp
class Solution {
public:
    long long numberOfRightTriangles(vector<vector<int>>& grid) {
        long long ans=0;
        int n=grid.size();
        int m=grid[0].size();

        vector<long long>row(n);
        vector<long long>col(m);
        for(int i=0;i<n;i++) //calculate 1s in each row
        {
            long long cnt=0;
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]==1)cnt++;
            }
            row[i]=cnt;
        }
        for(int i=0;i<m;i++)//calculate 1s in each column
        {
            long long cnt=0;
            for(int j=0;j<n;j++)
            {
                if(grid[j][i]==1)cnt++;
            }
            col[i]=cnt;
        }

        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]==1)
                ans+=(row[i]-1)*(col[j]-1); 
            }
        }
        return ans;
    }
};